src/avx512bw/avx512bw-from-valid-utf8.inl.cpp

```diff
@@ -0,0 +1,339 @@
+
+        //    t2 = 8 x [110110aaaaaaaaaa|110111bbbbbbbbbb] -- copy hi word from t1 to t0
+        //         0xba = (t2 and not v_fc00_fc000) or v_d800_dc00
+        const __m512i t3 = _mm512_ternarylogic_epi32(t2, v_fc00_fc00, v_d800_dc00, 0xba);
+
+        _mm512_storeu_si512((__m512i*)words, t3);
+    }
+
+    // 3. store valid 16-bit values
+    _mm256_mask_storeu_epi16((__m256i*)output, valid, _mm512_cvtepi32_epi16(utf32));
+
+    int sp = __builtin_popcount(sp_mask);
+
+    // 4. copy surrogate pairs
+    uint32_t mask = sp_mask;
+    for (int i=count; i >= 0 && mask != 0; i--) {
+        if (test_and_clear_bit(mask, i)) {
+            output[i + sp] = words[2*i + 0];
+            sp -= 1;
+            output[i + sp] = words[2*i + 1];
+        } else {
+            output[i + sp] = output[i];
+        }
+    }
+
+    return count + __builtin_popcount(sp_mask);
+}
+
+/*
+    valid_utf8_to_fixed_length converts a valid UTF-8 string into UTF-32.
+
+    The `OUTPUT` template type decides what to do with UTF-32: store
+    it directly or convert into UTF-16 (with AVX512).
+
+    Input:
+    - str           - valid UTF-8 string
+    - len           - string length
+    - out_buffer    - output buffer
+
+    Result:
+    - pair.first    - the first unprocessed input byte
+    - pair.second   - the first unprocessed output word
+*/
+
+template <typename OUTPUT>
+std::pair<const char*, OUTPUT*> valid_utf8_to_fixed_length(const char* str, size_t len, OUTPUT* out_buffer) {
+    constexpr bool UTF32 = std::is_same<OUTPUT, uint32_t>::value;
+    constexpr bool UTF16 = std::is_same<OUTPUT, char16_t>::value;
+    static_assert(UTF32 or UTF16, "output type has to be uint32_t (for UTF-32) or char16_t (for UTF-16)");
+
+    const char* ptr = str;
+    const char* end = ptr + len;
+
+    OUTPUT* output = out_buffer;
+
+    while (ptr + 16 + 4 < end) {
+        // 1. Load all possible 4-byte substring into an AVX512 register.
+        const __m512i input = expand_bytes(ptr);
+
+        /* 2. Classify which words contain valid UTF-8 characters.
+              We test if the 0th byte is not a continuation byte (0b10xxxxxx) */
+        __mmask16 valid;
+        {
+            const __m512i t0 = _mm512_and_si512(input, v_0000_00c0);
+            valid = _mm512_cmpneq_epu32_mask(t0, v_0000_0080);
+        }
+        const int valid_count = __builtin_popcount(valid);
+
+        // 3. Find out character classes
+        __m512i char_class;
+        {
+            const __m512i t0 = _mm512_srli_epi32(input, 4);
+            // char_class = ((input >> 4) & 0x0f) | 0x80808000
+            char_class = _mm512_ternarylogic_epi32(t0, v_0000_000f, v_8080_8000, 0xea);
+        }
+
+        // 3. Convert words into UTF-32
+        const __m512i utf32 = expanded_utf8_to_utf32(char_class, input);
+
+        // 4. Pack only the valid words
+        const __m512i out = _mm512_mask_compress_epi32(_mm512_setzero_si512(), valid, utf32);
```

Comment:
That's a key trick. Obviously, _mm512_mask_compress_epi32 is not going to come for free. It is 6 cycles of latency and it has a soso throughput,  but it plays an important role.